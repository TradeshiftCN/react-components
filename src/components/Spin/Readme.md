### Basic Usage

```jsx
<Spin size="small"/>
<Spin />
<Spin size="large"/>
```

### Embedded mode

```jsx
initialState = { spinning: false, blocking: false };
<div>
	<Spin
		spinning={state.spinning}
		blocking={state.blocking}
		tip="Loading..."
		style={{
			border: '1px solid red'
		}}
	>
		<p>content...</p>
		<p>content...</p>
		<p>content...</p>
		<p>content...</p>
		<p>content...</p>
		<p>content...</p>
		<p>content...</p>
	</Spin>
	<label style={{ lineHeight: '30px' }}>
		Loading state:
		<input
			type="checkbox"
			checked={state.spinning}
			onChange={e => setState({ spinning: e.target.checked })}
		/>
	</label>
	<br />
	<label style={{ lineHeight: '30px' }}>
		Blocking state:
		<input
			type="checkbox"
			checked={state.blocking}
			onChange={e => setState({ blocking: e.target.checked })}
		/>
	</label>
</div>;
```
