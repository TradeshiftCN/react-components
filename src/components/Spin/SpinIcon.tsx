import * as React from 'react';
import cx from 'classnames';

export interface SpinIconProps {
	size: number;
	prefixCls: string;
	tip?: string;
	className?: string;
	style?: React.CSSProperties;
}

export default function SpinIcon(props: SpinIconProps) {
	const strips = [];
	const { size } = props;
	const stripWidth = size * 0.28;
	const stripHeight = size * 0.06;
	const duration = 0.08 * 12;
	for (let i = 0; i < 12; i++) {
		strips.push(
			<div
				className={`${props.prefixCls}__strip`}
				key={i}
				style={{
					width: stripWidth,
					height: stripHeight,
					transformOrigin: `${size / 2}px 0px`,
					transform: `rotate(${i * 30}deg) translateY(-50%)`,
					animationDelay: `${0.08 * i}s`,
					animationDuration: `${duration}s`
				}}
			/>
		);
	}
	return (
		<div className={cx(props.prefixCls, props.className)} style={props.style}>
			<div
				className={`${props.prefixCls}__strips`}
				style={{
					width: size,
					height: size
				}}
			>
				{strips}
			</div>
			{props.tip && <div className={`${props.prefixCls}__tip`}>{props.tip}</div>}
		</div>
	);
}
