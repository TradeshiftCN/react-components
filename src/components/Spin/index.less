@import '../../style/variables/index';

@keyframes spinFadeIn {
	0% {
		opacity: 0.5;
	}
	25% {
		opacity: 0.9;
	}
	100% {
		opacity: 0.5;
	}
}
.ts-rc-spin {
	display: inline-block;
	position: relative;
	&__strip {
		position: absolute;
		left: 0;
		top: 50%;
		background-color: rgb(85, 85, 85);
		box-shadow: rgba(0, 0, 0, 0.1) 0px 0px 1px;
		animation: spinFadeIn linear infinite;
		border-radius: ~'10.7% / 50%';
		opacity: 0.5;
	}
	&__container {
		position: relative;
	}
	&__spinning {
		position: absolute;
		left: 50%;
		top: 50%;
		transform: translate(-50%, -50%);
		font-size: 0px;
	}

	&__mask {
		position: absolute;
		top: 0;
		right: 0;
		bottom: 0;
		left: 0;
		background-color: rgba(0, 0, 0, 0.71);
		& + .ts-rc-spin__spinning {
			.ts-rc-spin__strip {
				background-color: @ts-color-white;
			}
			.ts-rc-spin__tip {
				color: @ts-color-white;
			}
		}
	}

	&__tip {
		width: 100%;
		text-align: center;
		position: absolute;
		top: 50%;
		font-size: @ts-fontsize;
	}
}
