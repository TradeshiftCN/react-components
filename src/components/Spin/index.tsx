import React, { Component } from 'react';
import cx from 'classnames';
import SpinIcon from './SpinIcon';
import _ from 'lodash';
import PropTypes from 'prop-types';
import { Overwrite } from '../../helper';

export type SpinProps = Overwrite<
	React.HTMLProps<HTMLDivElement>,
	{
		spinning?: boolean;
		size?: 'small' | 'default' | 'large';
		blocking?: boolean;
		tip?: string;
	}
>;
export type SpinState = {};

const prefixCls = 'ts-rc-spin';
const sizes = {
	small: 30,
	default: 65,
	large: 100
};
class Spin extends Component<SpinProps, SpinState> {
	static propTypes = {
		spinning: PropTypes.bool,
		size: PropTypes.oneOf(['small', 'default', 'large']),
		blocking: PropTypes.bool,
		tip: PropTypes.string
	};
	static defaultProps = {
		size: 'default',
		spinning: false,
		blocking: false
	};
	render() {
		const spinIconProps = {
			prefixCls: prefixCls,
			size: sizes[this.props.size!],
			tip: this.props.tip
		};
		if (!React.Children.count(this.props.children)) {
			return (
				<SpinIcon {...spinIconProps} className={this.props.className} style={this.props.style} />
			);
		}
		const containerProps = _.omit(this.props, ['className', 'spinning', 'size', 'tip', 'blocking']);
		return (
			<div className={cx(`${prefixCls}__container`, this.props.className)} {...containerProps}>
				{this.props.children}
				{this.props.spinning && this.props.blocking && <div className={`${prefixCls}__mask`} />}
				{this.props.spinning && (
					<div className={`${prefixCls}__spinning`}>
						<SpinIcon {...spinIconProps} />
					</div>
				)}
			</div>
		);
	}
}

export default Spin;
