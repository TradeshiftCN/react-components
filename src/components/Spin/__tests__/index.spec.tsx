import React from 'react';
import { shallow } from 'enzyme';
import Spin from '..';
import SpinIcon from '../SpinIcon';

describe('Spin', () => {
	it('render', () => {
		const wrapper = shallow(<Spin />);

		expect(wrapper).toMatchSnapshot();
	});

	it('embed', () => {
		const wrapper = shallow(<Spin>text</Spin>);
		expect(wrapper).toMatchSnapshot();
		wrapper.setProps({ spinning: true });
		expect(wrapper).toMatchSnapshot();
		wrapper.setProps({ blocking: true });
		expect(wrapper).toMatchSnapshot();
	});

	it('tip', () => {
		expect(shallow(<Spin tip="tip" />)).toMatchSnapshot();
		expect(
			shallow(
				<Spin tip="tip" spinning>
					text
				</Spin>
			)
		).toMatchSnapshot();
	});

	it('icon', () => {
		const wrapper = shallow(<SpinIcon size={100} prefixCls="test" />);
		expect(wrapper).toMatchSnapshot();
		wrapper.setProps({ tip: 'tip' });

		expect(wrapper.find('.test__tip')).toMatchSnapshot();
	});
});
