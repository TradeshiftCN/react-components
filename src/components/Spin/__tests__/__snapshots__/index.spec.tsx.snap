// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spin embed 1`] = `
<div
  blocking={false}
  className="ts-rc-spin__container"
>
  text
</div>
`;

exports[`Spin embed 2`] = `
<div
  blocking={false}
  className="ts-rc-spin__container"
>
  text
  <div
    className="ts-rc-spin__spinning"
  >
    <SpinIcon
      prefixCls="ts-rc-spin"
      size={65}
    />
  </div>
</div>
`;

exports[`Spin embed 3`] = `
<div
  blocking={true}
  className="ts-rc-spin__container"
>
  text
  <div
    className="ts-rc-spin__mask"
  />
  <div
    className="ts-rc-spin__spinning"
  >
    <SpinIcon
      prefixCls="ts-rc-spin"
      size={65}
    />
  </div>
</div>
`;

exports[`Spin icon 1`] = `
<div
  className="test"
>
  <div
    style={
      Object {
        "height": 100,
        "width": 100,
      }
    }
  >
    <div
      className="test__strip"
      key="0"
      style={
        Object {
          "animationDelay": "0s",
          "animationDuration": "0.96s",
          "height": 6,
          "transform": "rotate(0deg) translateY(-50%)",
          "transformOrigin": "50px 0px",
          "width": 28.000000000000004,
        }
      }
    />
    <div
      className="test__strip"
      key="1"
      style={
        Object {
          "animationDelay": "0.08s",
          "animationDuration": "0.96s",
          "height": 6,
          "transform": "rotate(30deg) translateY(-50%)",
          "transformOrigin": "50px 0px",
          "width": 28.000000000000004,
        }
      }
    />
    <div
      className="test__strip"
      key="2"
      style={
        Object {
          "animationDelay": "0.16s",
          "animationDuration": "0.96s",
          "height": 6,
          "transform": "rotate(60deg) translateY(-50%)",
          "transformOrigin": "50px 0px",
          "width": 28.000000000000004,
        }
      }
    />
    <div
      className="test__strip"
      key="3"
      style={
        Object {
          "animationDelay": "0.24s",
          "animationDuration": "0.96s",
          "height": 6,
          "transform": "rotate(90deg) translateY(-50%)",
          "transformOrigin": "50px 0px",
          "width": 28.000000000000004,
        }
      }
    />
    <div
      className="test__strip"
      key="4"
      style={
        Object {
          "animationDelay": "0.32s",
          "animationDuration": "0.96s",
          "height": 6,
          "transform": "rotate(120deg) translateY(-50%)",
          "transformOrigin": "50px 0px",
          "width": 28.000000000000004,
        }
      }
    />
    <div
      className="test__strip"
      key="5"
      style={
        Object {
          "animationDelay": "0.4s",
          "animationDuration": "0.96s",
          "height": 6,
          "transform": "rotate(150deg) translateY(-50%)",
          "transformOrigin": "50px 0px",
          "width": 28.000000000000004,
        }
      }
    />
    <div
      className="test__strip"
      key="6"
      style={
        Object {
          "animationDelay": "0.48s",
          "animationDuration": "0.96s",
          "height": 6,
          "transform": "rotate(180deg) translateY(-50%)",
          "transformOrigin": "50px 0px",
          "width": 28.000000000000004,
        }
      }
    />
    <div
      className="test__strip"
      key="7"
      style={
        Object {
          "animationDelay": "0.56s",
          "animationDuration": "0.96s",
          "height": 6,
          "transform": "rotate(210deg) translateY(-50%)",
          "transformOrigin": "50px 0px",
          "width": 28.000000000000004,
        }
      }
    />
    <div
      className="test__strip"
      key="8"
      style={
        Object {
          "animationDelay": "0.64s",
          "animationDuration": "0.96s",
          "height": 6,
          "transform": "rotate(240deg) translateY(-50%)",
          "transformOrigin": "50px 0px",
          "width": 28.000000000000004,
        }
      }
    />
    <div
      className="test__strip"
      key="9"
      style={
        Object {
          "animationDelay": "0.72s",
          "animationDuration": "0.96s",
          "height": 6,
          "transform": "rotate(270deg) translateY(-50%)",
          "transformOrigin": "50px 0px",
          "width": 28.000000000000004,
        }
      }
    />
    <div
      className="test__strip"
      key="10"
      style={
        Object {
          "animationDelay": "0.8s",
          "animationDuration": "0.96s",
          "height": 6,
          "transform": "rotate(300deg) translateY(-50%)",
          "transformOrigin": "50px 0px",
          "width": 28.000000000000004,
        }
      }
    />
    <div
      className="test__strip"
      key="11"
      style={
        Object {
          "animationDelay": "0.88s",
          "animationDuration": "0.96s",
          "height": 6,
          "transform": "rotate(330deg) translateY(-50%)",
          "transformOrigin": "50px 0px",
          "width": 28.000000000000004,
        }
      }
    />
  </div>
</div>
`;

exports[`Spin icon 2`] = `
<div
  className="test__tip"
  style={
    Object {
      "marginTop": 58,
    }
  }
>
  tip
</div>
`;

exports[`Spin render 1`] = `
<SpinIcon
  prefixCls="ts-rc-spin"
  size={65}
/>
`;

exports[`Spin tip 1`] = `
<SpinIcon
  prefixCls="ts-rc-spin"
  size={65}
  tip="tip"
/>
`;

exports[`Spin tip 2`] = `
<div
  blocking={false}
  className="ts-rc-spin__container"
>
  text
  <div
    className="ts-rc-spin__spinning"
  >
    <SpinIcon
      prefixCls="ts-rc-spin"
      size={65}
      tip="tip"
    />
  </div>
</div>
`;
